File: tries/421.py
# Microsoft
# Bit Trie(XOR) -> left for 0 and right for 1
# Bit Trie max height -> log(num)
# Iterating bit trie -> n * log(max no) 
# Tc = n * log(max no) [32 bit integer, so max height 32 too] => Tc = n * log(32) => O(n*32) => O(n)
# Sc is like Tc also
class TrieNode:
    def __init__(self):
        self.left = None
        self.right = None
    

class Solution: 
    def insert(self, root, num):
        pCrawl = root
        for i in range(31, -1, -1):
            ith_bit = (num >> i) & 1
            if ith_bit == 0:
                if pCrawl.left is None:
                    pCrawl.left = TrieNode()
                pCrawl = pCrawl.left
            else:
                if pCrawl.right is None:
                    pCrawl.right = TrieNode()
                pCrawl = pCrawl.right
    
    def findMaxXor(self, root, num):
        maxXor = 0
        pCrawl = root
        for i in range(31, -1, -1):
            ith_ibt = (num >> i) & 1
            if ith_ibt == 1:
                if pCrawl.left is not None:
                    maxXor += (1 << i)
                    pCrawl = pCrawl.left
                else: 
                    pCrawl = pCrawl.right
            else:
                if pCrawl.right is not None:
                    max += (1 << i)
                    pCrawl = pCrawl.right
                else:
                    pCrawl = pCrawl.left
        return maxXor

    def findMaximumXor(self, nums):
        root = TrieNode()
        for num in nums:
            self.insert(root, num)
        maxResult = 0
        for num in nums:
            temp = self.findMaxXor(root, num)
            maxResult = max(maxResult, temp)
        return maxResult

File: tries/421.js
class TrieNode {
    constructor() {
        this.left = null;
        this.right = null;
    }
}

class Solution {
    insert(root, num) {
        let pCrawl = root;
        for (let i = 31; i >= 0; i--) {
            const ith_bit = (num >> i) & 1;
            if (ith_bit === 0) {
                if (pCrawl.left === null) {
                    pCrawl.left = new TrieNode();
                }
                pCrawl = pCrawl.left;
            } else {
                if (pCrawl.right === null) {
                    pCrawl.right = new TrieNode();
                }
                pCrawl = pCrawl.right;
            }
        }
    }

    findMaxXor(root, num) {
        let maxXor = 0;
        let pCrawl = root;
        for (let i = 31; i >= 0; i--) {
            const ith_bit = (num >> i) & 1;
            if (ith_bit === 1) {
                if (pCrawl.left !== null) {
                    maxXor += (1 << i);
                    pCrawl = pCrawl.left;
                } else {
                    pCrawl = pCrawl.right;
                }
            } else {
                if (pCrawl.right !== null) {
                    maxXor += (1 << i);
                    pCrawl = pCrawl.right;
                } else {
                    pCrawl = pCrawl.left;
                }
            }
        }
        return maxXor;
    }

    findMaximumXor(nums) {
        const root = new TrieNode();
        for (const num of nums) {
            this.insert(root, num);
        }
        let maxResult = 0;
        for (const num of nums) {
            const temp = this.findMaxXor(root, num);
            maxResult = Math.max(maxResult, temp);
        }
        return maxResult;
    }
}